File: ssDNA_annotator/modules/plot_tree.R
#!/usr/bin/env Rscript


# Load required libraries
suppressPackageStartupMessages({
  library(treeio)
  library(ggtree)
  library(ggtreeExtra)
  library(ggplot2)
  library(dplyr)
  library(tidyr)
  library(ape)
  library(rlang)
})

# --- Parse command-line arguments using commandArgs() ---

args <- commandArgs(trailingOnly = TRUE)

# Set default options in a list
opt <- list(
  tree = NULL,
  outdir = NULL,
  metadata_1 = NULL,
  metadata_2 = NULL,
  alignment = NULL,
  layout = "rectangular",
  branch_length = "branch.length",
  open_angle = 0,
  offset = 0.14,
  tip_label = "family",
  color = FALSE,
  fig_width = 7,
  fig_height = 7,
  plot_tips = TRUE,
  use_dist_matrix = FALSE,
  plot_name = "tree_plot.pdf"
)

# Helper: parse arguments in the form --option=value
for (arg in args) {
  if (grepl("^--", arg)) {
    splitArg <- strsplit(sub("^--", "", arg), "=")[[1]]
    if (length(splitArg) == 2) {
      opt[[splitArg[1]]] <- splitArg[2]
    } else {
      # If flag provided without "=value", set to TRUE
      opt[[splitArg[1]]] <- TRUE
    }
  }
}

# Convert numeric arguments
opt$open_angle <- as.numeric(opt$open_angle)
opt$offset <- as.numeric(opt$offset)
opt$fig_width <- as.numeric(opt$fig_width)
opt$fig_height <- as.numeric(opt$fig_height)

# Convert boolean arguments (if provided as strings)
opt$color <- tolower(as.character(opt$color)) %in% c("true", "1", "yes")
opt$plot_tips <- tolower(as.character(opt$plot_tips)) %in% c("true", "1", "yes")
opt$use_dist_matrix <- tolower(as.character(opt$use_dist_matrix)) %in% c("true", "1", "yes")

# Validate required arguments
if (is.null(opt$tree)) {
  stop("Error: You must provide a tree file using --tree=FILE", call.=FALSE)
}
if (is.null(opt$outdir)) {
  stop("Error: You must provide an output directory using --outdir=DIR", call.=FALSE)
}

# --- Process the tree input ---

if (opt$use_dist_matrix) {
  # Assume the distance matrix file is the tree file name with a .mldist extension
  dist_file <- sub("\\.treefile$", ".mldist", opt$tree)
  message("Reading distance matrix from: ", dist_file)
  mldist_matrix <- read.table(dist_file, skip = 1, header = FALSE, row.names = 1)
  mldist_matrix <- as.dist(mldist_matrix)
  tree <- ape::bionj(mldist_matrix)
} else {
  tree <- read.tree(opt$tree)
}

# --- Merge tree with metadata if provided ---
if (!is.null(opt$metadata_2)) {
  # Both metadata files are provided.
  meta_data_1 <- read.csv(opt$metadata_1)
  meta_data_2 <- read.delim(opt$metadata_2, sep="\t")
  
  meta_data_2 <- meta_data_2 %>% rename(protein_description = Original_Name)
  
  # Merge using protein_description as key, then rename the join column to "label"
  meta_final <- left_join(meta_data_2, meta_data_1, by = "protein_description") %>%
                rename(label = Sanitized_Name)
  
  trda <- tree %>% left_join(meta_final, by = "label")
  
} else if (!is.null(opt$metadata_1)) {
  # Only one metadata file is provided.
  meta_data_1 <- read.csv(opt$metadata_1)
  
  meta_final <- meta_data_1 %>% rename(label = protein_description)
  trda <- tree %>% left_join(meta_final, by = "label")
  
} else {
  trda <- tree
}

# --- Check if the join was successful ---
# Here we assume that if metadata was provided, the 'tip_label' column is expected.
expected_column <- opt$tip_label
cols_to_check = trda %>% select(everything()) |> colnames()

if (!is.null(opt$metadata_1) | !is.null(opt$metadata_2)) {
    if (!(expected_column %in% cols_to_check)) {
    warning(paste("\n\nAfter joining, the expected column", expected_column, 
              "was not found. Please verify that the metadata values match the tree labels.\n\n"))
  }
}

if (!is.null(opt$metadata_1) | !is.null(opt$metadata_2)) {
  missing_matches <- trda %>% select(all_of(expected_column), label) %>% drop_na(all_of(expected_column))
  if(nrow(missing_matches) == 0) {
    warning(paste("\n\nTree tip labels do not have matching metadata in the", 
              expected_column, "column\n\n
              The plot may not be colored adequately.\n\n"))
  }
}


# --- Set plotting parameters ---
layout <- opt$layout
branch_length <- opt$branch_length
open_angle <- opt$open_angle
offset <- opt$offset
tip_label <- opt$tip_label

# --- Create base tree plot ---
if (layout %in% c("circular", "unrooted")) {
  p <- ggtree(trda,
              layout = layout,
              open.angle = open_angle,
              branch.length = branch_length,
              ladderize = TRUE,
              right = FALSE,
              root.position = 0,
              hang = 0.1) +
       theme_tree2()
  if (opt$plot_tips) {
    p <- p + geom_tiplab(offset = offset, aes(angle = angle))
  }
  p <- p + geom_tippoint(mapping = aes(shape = !!sym(tip_label), color = !!sym(tip_label)))
} else {
  p <- ggtree(trda,
              layout = layout,
              open.angle = open_angle,
              branch.length = branch_length,
              ladderize = TRUE,
              right = FALSE,
              root.position = 0,
              hang = 0.1) +
       theme_tree2()
  if (opt$plot_tips) {
    p <- p + geom_tiplab(offset = offset)
  }
  p <- p + geom_tippoint(mapping = aes(shape = !!sym(tip_label), color = !!sym(tip_label)))
}

# --- Add alignment plot if alignment file is provided ---
if (!is.null(opt$alignment)) {
  p <- msaplot(p, fasta = opt$alignment) + theme(legend.position = "none")
}

# --- Optionally color the tree if --color is TRUE and metadata is provided ---
if (opt$color && !is.null(opt$metadata_1) && !is.null(opt$metadata_2)) {
  new_labels <- trda %>% select(label, !!as.name(tip_label), isTip)
  grp <- split(new_labels$label[new_labels$isTip], new_labels[[tip_label]][new_labels$isTip])
  p <- groupOTU(p, grp, tip_label) +
      aes(color = !!sym(tip_label)) +
      theme(legend.position = "right")
}

# --- Save the plot ---
output_file <- file.path(opt$outdir, opt$plot_name)
message("Saving figure to: ", output_file)
ggsave(filename = output_file, plot = p, width = opt$fig_width, height = opt$fig_height)

# ./ssDNA_annotator/modules/plot_tree.R --tree=mytree.treefile --outdir=output_dir \
#                   --metadata_1=meta.csv --metadata_2=metadata_2.tsv \
#                   --alignment=alignment.fasta \
#                   --layout=circular --branch_length=branch.length \
#                   --open_angle=90 --offset=0.2 --tip_label=family 
#                   --color=TRUE --fig_width=8 --fig_height=8 --plot_tips=TRUE --use_dist_matrix=FALSE \
#                   --plot_name=my_custom_tree.pdf

File: cressent_core/modules/plot_tree.py
#!/usr/bin/env python3
import os
import sys
import subprocess
import argparse
import logging
from pathlib import Path

def setup_logging(output_dir):
    log_file = os.path.join(output_dir, "plot_tree.log")
    logging.basicConfig(
        level=logging.INFO,
        format='%(asctime)s - %(levelname)s - %(message)s',
        handlers=[
            logging.FileHandler(log_file, mode='w'),
            logging.StreamHandler()
        ]
    )

def main():
    # Define command-line arguments for the Python wrapper.
    parser = argparse.ArgumentParser(
        description="Wrapper to run the plot_tree_cli.R script from the ssDNA tool."
    )
    parser.add_argument("-t", "--tree", help="Input tree file (Newick format)")
    parser.add_argument("--dist_matrix", help="Use distance matrix method for tree construction")
    parser.add_argument("-o", "--outdir", required=True, help="Output directory to save the figure")
    parser.add_argument("--metadata_1", help="Optional CSV metadata file")
    parser.add_argument("--metadata_2", help="Optional TSV name table file")
    parser.add_argument("--alignment", help="Optional alignment file (FASTA) to include in the plot")
    parser.add_argument("--layout", default="rectangular", help="Tree layout (e.g., rectangular, circular, unrooted)")
    parser.add_argument("--branch_length", default="branch.length", help="Branch length parameter for ggtree")
    parser.add_argument("--open_angle", default=0, type=float, help="Open angle for circular/unrooted layouts")
    parser.add_argument("--offset", default=0.14, type=float, help="Tip label offset")
    parser.add_argument("--tip_label", default="family", help="Column name to use as the tip label")
    parser.add_argument("--color", default=True, type=lambda s: s.lower() in ['true', '1', 'yes'], help="Color tree by group (requires metadata)")
    parser.add_argument("--fig_width", type=float, default=7, help="Figure width (ggsave)")
    parser.add_argument("--fig_height", type=float, default=7, help="Figure height (ggsave)")
    parser.add_argument("--plot_tips", default=True, type=lambda s: s.lower() in ['true', '1', 'yes'], help="Include tip labels in the plot")
    parser.add_argument("--plot_name", default="tree_plot.pdf", help="Name of the output plot file")
    args = parser.parse_args()
    
    # Determine the absolute path of the modules directory.
    # Assuming the directory structure:
    # ssDNA_tool/
    #   ssDNA_annotator/
    #     modules/
    #       plot_tree.R
    modules_dir = Path(__file__).resolve().parent  # Directory where the script is located
    r_script_path = modules_dir / "plot_tree.R"  # Update with the actual path
    
    # Set up logging with log file in the specified output directory.
    setup_logging(args.outdir)
    logger = logging.getLogger("plot_tree.log")

    if not os.path.isfile(r_script_path):
        logger.error("R script 'plot_tree.R' not found in modules directory: %s", r_script_path)
        sys.exit(1)
    
    # Build a list of arguments to pass to the R script in the form --option=value
    r_args = []
    if args.tree is not None:
        r_args.append(f"--tree={args.tree}")
    if args.dist_matrix is not None:
        r_args.append(f"--dist_matrix={args.dist_matrix}")
    r_args.append(f"--outdir={args.outdir}")
    r_args.append(f"--layout={args.layout}")
    r_args.append(f"--branch_length={args.branch_length}")
    r_args.append(f"--open_angle={args.open_angle}")
    r_args.append(f"--offset={args.offset}")
    r_args.append(f"--tip_label={args.tip_label}")
    r_args.append(f"--fig_width={args.fig_width}")
    r_args.append(f"--fig_height={args.fig_height}")
    r_args.append(f"--plot_name={args.plot_name}")
    r_args.append(f"--color={str(args.color).upper()}")
    r_args.append(f"--plot_tips={str(args.plot_tips).upper()}")
    
    # Add optional arguments if provided
    if args.metadata_1:
        r_args.append(f"--metadata_1={args.metadata_1}")
    if args.metadata_2:
        r_args.append(f"--metadata_2={args.metadata_2}")
    if args.alignment:
        r_args.append(f"--alignment={args.alignment}")
    
    # Construct the full command: Rscript path/to/plot_tree_cli.R <arguments>
    command = ["Rscript", r_script_path] + r_args
    logger.info("Executing command:")
    logger.info(" ".join(map(str, command)))
    
    # Execute the R script with the specified arguments
    try:
        subprocess.run(command, check=True)
        logger.info("R script executed successfully.")
    except subprocess.CalledProcessError as e:
        logger.error("R script execution failed with error: %s", e)
        sys.exit(1)

if __name__ == "__main__":
    main()



# python /fs/project/PAS1117/ricardo/ssDNA_tool/ssDNA_annotator/modules/plot_tree.py \
#      --tree=mytree.treefile --outdir=output_dir \
#     --metadata=meta.csv --name_table=name_table.tsv --alignment=alignment.fasta \
#     --layout=circular --branch_length=branch.length --open_angle=90 --offset=0.2 \
#     --tip_label=family --color --fig_width=8 --fig_height=8 --plot_tips --dist_matrix \
#     --plot_name=my_custom_tree.pdf